### PR TITLE
Platform/david navapbc/15761/2/migrate queue message to lib

### PR DESCRIPTION
**Pull request by david-navapbc**
## THIS IS A WORK IN PROGRESS 

The test classes need to be updated to fix broken references to things that are now in the shared module 

The production code compiles

### CODE REVIEW

File: prime-router/src/main/kotlin/Report.kt [hunk 1]
```diff
@@ -60,56 +63,56 @@ const val REPORT_MAX_ERRORS = 100
...
-            } catch (ex: IllegalArgumentException) {
-                val msg = "$input is not a valid Option. Valid options: ${Options.activeValues.joinToString()}"
-                throw InvalidOptionException(msg)
-            }
-    }
-}
-
-annotation class OptionDeprecated
+//// options are used to process and route the report
+//enum class Options {
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/Report.kt]: 👍  yupyup - I'll make sure all the commented out stuff is gone before I remove the "draft" label

File: prime-router/src/main/kotlin/azure/ActionHistory.kt [hunk 2]
```diff
@@ -464,10 +465,10 @@ class ActionHistory(
         // TODO: Need to update this process to have a better way to determine what messages should be sent
         //  automatically as part of queueMessages and what are being send manually as part of the parent function.
         //  The automatic queueing uses the action name as the queue name, and this is not the case for FHIR actions
-        if (event.eventAction != Event.EventAction.BATCH &&
-            event.eventAction != Event.EventAction.DESTINATION_FILTER &&
-            event.eventAction != Event.EventAction.RECEIVER_FILTER &&
-            event.eventAction != Event.EventAction.TRANSLATE
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/azure/ActionHistory.kt:470]: the simple statement a lot of what's in RS is tightly coupled with the rest of RS in various ways. It was a challenge to find a place where I could cleave out the logic that went into "shared" without dragging a ton of other things along with it. 

EventAction is used by the queue logic that got ported over but Event is not. Moreover, if we migrate Event over we also have include jackson and jooq dependencies as well as other packages like 

```
gov.cdc.prime.router.fhirengine.engine.LookupTableValueSet
org.jooq.Catalog
org.jooq.EnumType
org.jooq.Schema
```

as well as other dependencies and packages because a lot of what's in RS has some kind of connection to other things in RS.

for example, `LookupTableValueSet` brings in 

```
gov.cdc.prime.router.Metadata
gov.cdc.prime.router.fhirengine.translation.hl7.SchemaException
gov.cdc.prime.router.fhirengine.translation.hl7.ValueSetCollection
```

bottom line - it's good we're doing the work to decouple the app and it's going to be challenging to do so cleanly along business logic lines at first until we get more of the app decoupled from itself 😅

File: prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt [hunk 3]
```diff
@@ -154,25 +150,25 @@ class FHIRFunctions(
...
         )
         // initialize the json types in PrimeRouterQueueMessage
-        initializeQueueMessages()
+        QueueMessage.ObjectMapperProvider.init()
 
         return when (val queueMessage = QueueMessage.deserialize(message)) {
             is QueueMessage.ReceiveQueueMessage -> {
-                FhirReceiveQueueMessage(
-                    queueMessage.reportId,
+                QueueMessage.ReceiveQueueMessage(
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/azure/FHIRFunctions.kt:162]: probably? it's all finally working so if you're ok with it I'd like to worry about integration-hell in getting current master and this branch in alignment before I remove this.

File: shared/src/main/kotlin/gov/cdc/prime/reportstream/shared/validation/IItemValidator.kt [hunk 4]
```diff
@@ -1,10 +1,9 @@
-package gov.cdc.prime.router.validation
+package gov.cdc.prime.reportstream.shared.validation
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply shared/src/main/kotlin/gov/cdc/prime/reportstream/shared/validation/IItemValidator.kt:1]: it had to be. I tried very hard to keep it in prime but it ended up dorking with business logic that assumed everything had some kind of validation logic associated with it - even if that logic was a noop placeholder.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIREngine.kt [hunk 5]
```diff
@@ -217,7 +219,7 @@ abstract class FHIREngine(
      *
      */
     fun run(
-        queueMessage: ReportPipelineMessage,
+        queueMessage: QueueMessage,
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIREngine.kt:222]: not sure what you mean?

File: prime-router/src/main/kotlin/fhirengine/engine/FHIREngine.kt [hunk 6]
```diff
@@ -231,9 +233,21 @@ abstract class FHIREngine(
...
+            // nothing here should be any other kind of queue message
+            // we need to check because interface QueueMessage does not have field 'reportId'
+            // we need the if (a) b else c call because ReportId can't be null and this makes the compiler a
+            // happy panda.
+            assert(queueMessage is QueueMessage.ReportInformation)
+            var reportId: ReportId =
+                if (queueMessage is QueueMessage.ReportInformation) {
+                    queueMessage.reportId
+                } else {
+                    UUID.randomUUID()
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIREngine.kt:245]: this was almost certainly code I ported over. I don't recall making a conscious decision to assign a rando uuid. looking at it this looks like one of the special handling situations you were saying in a prior comment may not be needed anymore. If that's the case I can look at getting rid of it.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRTranslator.kt [hunk 7]
```diff
@@ -243,22 +243,22 @@ class FHIRTranslator(
...
-    var encodedMsg = encode()
-    if (hasFiveEncodingChars) {
-        encodedMsg = encodedMsg.replace(defaultHl7EncodingFourChars, defaultHl7EncodingFiveChars)
-        // Set MSH-2 back in the in-memory message to preserve original value
-        Terser.set(msh, 2, 0, 1, 1, defaultHl7EncodingFiveChars)
-    }
-    return encodedMsg
-}
\ No newline at end of file
+// /**
```
[the remark this answers is not in the document]
Comment (david-navapbc) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRTranslator.kt:246]: ack - in all fairness though - this PR is marked draft for a reason 😛

File: settings.gradle.kts [hunk 8]
```diff
@@ -14,3 +14,33 @@ sourceControl {
 }
 rootProject.name = "prime-reportstream"
 include("shared", "submissions", "prime-router", "auth")
+
+dependencyResolutionManagement {
+    versionCatalogs {
+        create("libs") {
+            version("hapi-structures_caffeine_client", "7.2.2")
```
Comment (mkalish) [inline settings.gradle.kts:21]: No objection to doing it this way, but I think there is some overlap here with what we're doing in the `buildSrc` and the reportstream plugin.

That seems like a slightly different pattern for keeping depedencies in sync so it might be good to decide on

File: shared/build.gradle.kts [hunk 9]
```diff
@@ -12,7 +12,47 @@ repositories {
...
+
+//    implementation("gov.nist:hl7-v2-validation:1.6.5") {
+//        // These conflict with the javax.xml.transform package available in the base JDK and need to be excluded
+//        exclude("xerces")
+//        exclude("xml-apis")
+//    }
+
+    implementation(libs.hl7.v2.validation) {
+        // These conflict with the javax.xml.transform package available in the base JDK and need to be excluded
+        exclude("xerces")
```
Comment (mkalish) [inline shared/build.gradle.kts:44]: This might be a reason to go with the plugin approach to avoid having to do this in multiple projects

File: shared/src/main/kotlin/gov/cdc/prime/reportstream/shared/HL7MessageParseAndConvertConfiguration.kt
Comment (mkalish) [inline shared/src/main/kotlin/gov/cdc/prime/reportstream/shared/HL7MessageParseAndConvertConfiguration.kt:1]: This seems strange to have moved into shared since I don't think other places would ever need this?

**Review by mkalish — COMMENTED** [review]
Appreciate all the work that went into this!

Overall, I think this maybe needs to cook a little more?  It seems strange to me that pieces of pipeline moved to shared, especially ones that really aren't going to be used anywhere else.  When Brick and I were looking at the issue a while ago, we had discussed that the messages themselves really needed to by analyzed to see what data was required to be on them; we thought there was a a chance that:
- they could be substantially thinned down with more data fetched from the db (for example `FhirConvertQueueMessage` includes a `schemaName` property which should be fetched from the db)
  - another example of that could be topic which can be retrieved from also looking up the sender; that might help with why validation needed to move over
- there might not even need to be a need for subclasses of QueueMessage once the fields were thinned
- that the only need for a common interface was so that every message could be serialized the same and that might be better modeled as just a mixin that all the independent messages inherit

I also really like the version catalog idea and wonder how we can mesh it/provide guidance with the existing reportstream plugin.

**Comment by arnejduranovic** [comment]
No longer needed
